lib/edgehog_device_forwarder/tokens.ex

```diff
@@ -0,0 +1,67 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Tokens do
+  alias EdgehogDeviceForwarder.Tokens.Data
+  alias EdgehogDeviceForwarder.Monitor
+  require Logger
+
+  @tokens EdgehogDeviceForwarder.Cache.for(__MODULE__)
+  @type token :: String.t()
+  @type device_id :: Data.device_id()
+  @type realm :: Data.realm()
+  @type socket :: Data.socket()
+
+  @spec add_device_info(token, device_id, realm) ::
+          :ok | {:error, :token_not_found}
+  def add_device_info(token, device_id, realm) do
+    ConCache.update(@tokens, token, fn
+      nil -> {:error, :token_not_found}
+      token_data -> {:ok, Data.put_device_info(token_data, device_id, realm)}
+    end)
+  end
+
+  @spec reserve(token) :: :ok | {:error, :token_already_exists}
+  def reserve(token) do
+    with {:error, :already_exists} <- ConCache.insert_new(@tokens, token, :reserved) do
+      {:error, :token_already_exists}
+    end
+  end
+
+  @spec register(token, socket) :: :ok | {:error, :invalid_data}
+  def register(token, socket) do
+    ConCache.update(@tokens, token, fn
+      :reserved ->
+        start_monitor(socket, token)
+        {:ok, %Data{socket: socket}}
+
+      other ->
+        "Tokens.Data.register: token #{inspect(token)}: found #{inspect(other)}"
+        |> Logger.notice(tag: "token_data_error")
+
+        {:error, :invalid_data}
+    end)
+  end
+
+  @spec close(token) :: :ok
+  def close(token) do
+    ConCache.delete(@tokens, token)
+
+    "connection closed for token #{inspect(token)}"
+    |> Logger.info(tag: "device_event")
+  end
+
+  @spec fetch(token) :: {:ok, socket} | {:error, :token_not_found}
+  def fetch(token) do
+    case ConCache.get(@tokens, token) do
+      nil -> {:error, :token_not_found}
+      %Data{socket: socket} -> {:ok, socket}
+    end
+  end
+
+  @spec start_monitor(socket, token) :: :ok
+  defp start_monitor(socket, token) do
+    on_close = fn -> close(token) end
+    Monitor.monitor(socket, on_close)
```

Comment:
Personally, I'd write 
```
    Monitor.monitor(socket, fn -> close(token) end)
```
but de gustibus non disputandum est

Reply:
since the new (although not definitive) name for the monitor module/function is more descriptive, I felt more confident in putting the lambda directly in the function call as you suggested